File: java/com/google/copybara/git/github/api/GitHubApp.java
/*
 * Copyright (C) 2019 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git.github.api;

import com.google.api.client.util.Key;
import com.google.common.base.MoreObjects;
import javax.annotation.Nullable;
import net.starlark.java.annot.StarlarkBuiltin;
import net.starlark.java.annot.StarlarkMethod;
import net.starlark.java.eval.StarlarkValue;

/** Represents a GitHub App detail. https://developer.github.com/v3/apps/#response */
@StarlarkBuiltin(
    name = "github_app_obj",
    doc = "Detail about a GitHub App.")
public class GitHubApp implements StarlarkValue {

  @Key private int id;

  @Key private String slug;

  @Key private String name;

  @StarlarkMethod(name = "id", doc = "The GitHub App's Id", structField = true)
  public int getId() {
    return id;
  }

  @StarlarkMethod(
      name = "slug",
      doc = "The url-friendly name of the GitHub App.",
      structField = true
  )
  public String getSlug() {
    return slug;
  }

  @StarlarkMethod(
      name = "name",
      doc = "The GitHub App's name",
      structField = true,
      allowReturnNones = true)
  @Nullable
  public String getName() {
    return name;
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("id", id)
        .add("slug", slug)
        .add("name", name)
        .toString();
  }

}

File: src/Copybara.Core/Git/GitHub/Api/GitHubApp.cs
/*
 * Copyright (C) 2019 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Text.Json.Serialization;
using Starlark.Annot;
using Starlark.Eval;

namespace Copybara.Git.GitHub.Api;

/// <summary>Represents a GitHub App detail. https://developer.github.com/v3/apps/#response</summary>
[StarlarkBuiltin("github_app_obj", Doc = "Detail about a GitHub App.")]
public class GitHubApp : IStarlarkValue
{
    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("slug")]
    public string? Slug { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [StarlarkMethod("id", Doc = "The GitHub App's Id", StructField = true)]
    public int GetId() => Id;

    [StarlarkMethod("slug", Doc = "The url-friendly name of the GitHub App.", StructField = true)]
    public string? GetSlug() => Slug;

    [StarlarkMethod("name", Doc = "The GitHub App's name", StructField = true, AllowReturnNones = true)]
    public string? GetName() => Name;

    public override string ToString() => $"GitHubApp{{id={Id}, slug={Slug}, name={Name}}}";
}
